File: PHot100/middle/Solution15.py
class Solution15:
    def threeSum(self, nums):
        if len(nums)<3:
            return []

        nums.sort()
        if nums[0]>0 or nums[-1]<0:
            return []
        else:
            res = []
            i = 0
            while True:
                # boundary
                if i> len(nums)-3 or nums[i]>0:
                    break
                goal = 0-nums[i]
                # to find j,k in nums[i+1::] to make sum(j,k) = goal
                j = i+1
                k = len(nums)-1
                while True:
                    if j>=k:
                        break
                    if nums[j]+nums[k] == goal:
                        if [nums[i],nums[j],nums[k]] not in res:
                            res.append([nums[i],nums[j],nums[k]])

                    if nums[j]+nums[k] < goal:
                        j += 1
                    else:
                        k-=1
                i += 1
        return res


s = Solution15
nums = [-1,0,1,2,-1,-4]

print(s.threeSum(s,nums))

File: java_solution/Solution15.java
import org.junit.Test;

import java.util.*;
import java.util.stream.Collectors;

import static org.junit.Assert.*;

public class Solution15 {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> item = new ArrayList<>();
        res.add(item);
        /* nums.length <4 */
        if(nums.length<4){
            /* nums.length = 3*/
            if(nums.length == 3){
                if(nums[0]+ nums[1]+nums[2] == 0){
                    int [] pos = new int[]{1,2,3};
                    List<Integer> itemi = Arrays.stream(pos).boxed().collect(Collectors.toList());
                    res.add(itemi);
                }
            }
            return res;
        }

        /*       length>=4             */

        /* rank the nums*/
        Arrays.sort(nums);
        /* min>0 and max<0 reurn []*/
        if(nums[0]>0 || nums[nums.length-1]<0){
            return res;
        }

        /* nums.length>4 and min<=0 max>=0*/
        



        return null;
    }
    @Test
    public void TestSolution15(){
        int [] nums = new int[]{-1,0,1,2,-1,-4};
        List<List<Integer>> autual = threeSum(nums);

        nums = new int[]{};
        autual = threeSum(nums);

        nums = new int[]{0};
        autual = threeSum(nums);



    }

}
